aptos-move/block-executor/src/executor.rs

```diff
@@ -136,16 +136,32 @@ where
         // For tracking whether the recent execution wrote outside of the previous write/delta set.
         let mut updates_outside = false;
         let mut apply_updates = |output: &E::Output| {
-            // First, apply writes.
+            // TODO: This may seem like code duplication, but we will have three
+
+            // First, apply data writes.
             let write_version = (idx_to_execute, incarnation);
-            for (k, v) in output.get_writes().into_iter() {
+            for (k, v) in output.get_resource_writes().into_iter() {
+                if !prev_modified_keys.remove(&k) {
+                    updates_outside = true;
+                }
+                versioned_cache.write(k, write_version, v);
+            }
+
+            // Separately, record module writes.
+            for (k, v) in output.get_module_writes().into_iter() {
                 if !prev_modified_keys.remove(&k) {
                     updates_outside = true;
                 }
                 versioned_cache.write(k, write_version, v);
             }
 
-            // Then, apply deltas.
+            // Aggregators also require special handling: both rites and deltas go to the same map.
+            for (k, v) in output.get_aggregator_writes().into_iter() {
+                if !prev_modified_keys.remove(&k) {
+                    updates_outside = true;
+                }
+                versioned_cache.write(k, write_version, v);
+            }
```

Comment:
am I blind, I don't see any differences between these three loops? could we just chain the iterators and have one loop?

Reply:
That's correct, they are identical. We use the same mvhashmap right now, but there will be three -- one for each loop, so duplication will be removed very soon? I had a comment above about that. Let me know if you prefer chaining

Reply:
i thought we already expose a chaining functions, we can just use that for now